src/EditorFeatures/CSharpTest/RemoveUnnecessaryParentheses/RemoveUnnecessaryParenthesesTests.cs

```diff
@@ -0,0 +1,815 @@
+@"class C
+{
+    void M(C c1, C c2, C c3)
+    {
+        var x = $$(c1 + c2) + c3;
+    }
+
+    public static C operator +(C c1, C c2) => null;
+}",
+@"class C
+{
+    void M(C c1, C c2, C c3)
+    {
+        var x = c1 + c2 + c3;
+    }
+
+    public static C operator +(C c1, C c2) => null;
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestMissingForOverloadedOperatorOnRight()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M(C c1, C c2, C c3)
+    {
+        var x = c1 + $$(c2 + c3);
+    }
+
+    public static C operator +(C c1, C c2) => null;
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestShiftRequiredForClarity1()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        int x =  $$(1 + 2) << 3;
+    }
+}", parameters: new TestParameters(options: RequireShiftParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestShiftRequiredForClarity2()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        int x = $$(1 + 2) << 3;
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestRemoveShiftIfNotNecessary1()
+        {
+            await TestInRegularAndScript1Async(
+@"class C
+{
+    void M()
+    {
+        int x = $$(1 + 2) << 3;
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int x = 1 + 2 << 3;
+    }
+}", parameters: new TestParameters(options: RemoveAllUnnecessaryParentheses));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestRemoveCoalesceIfNotNecessary1()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        int x = $$(a ?? b) ?? c;
+    }
+}", parameters: new TestParameters(options: RemoveAllUnnecessaryParentheses));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestRemoveCoalesceIfNotNecessary2()
+        {
+            await TestInRegularAndScript1Async(
+@"class C
+{
+    void M()
+    {
+        int x = a ?? $$(b ?? c);
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int x = a ?? b ?? c;
+    }
+}", parameters: new TestParameters(options: RemoveAllUnnecessaryParentheses));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
```

Comment:
Nit: I'm not sure how explicit the IDE team is about this preference, but I would have put the `Trait` on the test class, as it gets inherited by the individual tests.

Reply:
This is the current IDE preference.  I would prefer we not change that unless we fix up all tests to have that form.